watcher/watcher.go

```diff
@@ -600,75 +588,208 @@ func (w *Watcher) checkState(
+		} else {
+			delta, shouldScaleUp, changedState, inSync, err = w.checkLegacyTrigger(autoScalingInfo, autoScalingInfo.Up, w.ScaleUpInfo, roomCount, scheduler, nowTimestamp)
+		}
+	}
 
-	isAboveThreshold, err := w.ScaleUpInfo.SendUsageAndReturnStatus(
-		w.SchedulerName,
-		capacity, roomCount.Occupied, roomCount.Total(),
-		threshold, usage,
-	)
-	if err != nil {
-		return false, false, false, err
+	if shouldScaleUp == false && inSync == true && roomCount.Total() > autoScalingInfo.Min {
+		// Down
+		if len(autoScalingInfo.Down.MetricsTrigger) > 0 {
+			delta, shouldScaleDown, changedState, inSync, err = w.checkMetricsTrigger(autoScalingInfo, autoScalingInfo.Down, w.ScaleDownInfo, roomCount, scheduler, nowTimestamp)
+		} else {
+			delta, shouldScaleDown, changedState, inSync, err = w.checkLegacyTrigger(autoScalingInfo, autoScalingInfo.Down, w.ScaleDownInfo, roomCount, scheduler, nowTimestamp)
+		}
+	} else if inSync == true { // always send current usage
+		if len(autoScalingInfo.Down.MetricsTrigger) > 0 {
+			for _, trigger := range autoScalingInfo.Down.MetricsTrigger {
+				w.ScaleDownInfo.SendUsage(
+					w.SchedulerName, trigger.Metric,
+					roomCount.Ready, roomCount.Total(),
+				)
+			}
+		} else {
+			w.ScaleDownInfo.SendUsage(
+				w.SchedulerName, "legacy",
+				roomCount.Ready, roomCount.Total(),
+			)
+		}
 	}
-	if isAboveThreshold {
-		inSync = false
-		if scheduler.State != models.StateSubdimensioned {
-			scheduler.State = models.StateSubdimensioned
-			scheduler.StateLastChangedAt = nowTimestamp
+
+	if inSync && scheduler.State != models.StateInSync {
+		scheduler.State = models.StateInSync
+		scheduler.StateLastChangedAt = nowTimestamp
+		changedState = true
+	}
+
+	return delta, shouldScaleUp, shouldScaleDown, changedState, err
+}
+
+func (w *Watcher) checkMetricsTrigger(
+	autoScalingInfo *models.AutoScaling,
+	scalingPolicy *models.ScalingPolicy,
+	scaleInfo *models.ScaleInfo,
+	roomCount *models.RoomsStatusCount,
+	scheduler *models.Scheduler,
+	nowTimestamp int64,
+) (int, bool, bool, bool, error) { // delta, shouldScaleUpOrDown, changedState, inSync
+	isDown := autoScalingInfo.Down == scalingPolicy
+	unbalancedState := models.StateSubdimensioned
+	shouldScaleUpOrDown := false
+	changedState := false
+	inSync := true
+	delta := 0
+
+	for _, trigger := range scalingPolicy.MetricsTrigger {
+		threshold := trigger.Threshold
+		usage := float32(trigger.Usage) / 100
+		capacity := trigger.Time / w.AutoScalingPeriod
+		point := roomCount.Occupied
+		unbalancedState = models.StateSubdimensioned
+
+		if isDown {
+			usage = 1 - usage
+			point = roomCount.Ready
+			unbalancedState = models.StateOverdimensioned
+		}
+
+		// If usage is above limit, scale up.
+		if isDown == false && 100*roomCount.Occupied >= roomCount.Total()*trigger.Limit {
+			w.Logger.Debug("Usage is above limit. Should scale up")
```

Comment:
This can be extracted to another function. The idea is reduce the size of this function and simplify it